### ISSUE TITLE
methodOptions, code, locate don't work on MethodFunctionSingle

### ISSUE THREAD

**Issue by pzinn** [Macaulay2/M2#2758]
maybe I'm misunderstanding something, but `code html`, `locate html`, `methodOptions html` don't produce anything, for example. Easily fixable, hopefully?

**Comment by mahrud** [comment]
That's how those tools work on any method function; try `locate methods html` instead.

**Comment by pzinn** [comment]
well, `methodOptions` should definitely work no matter what. As to `locate`, it does work on `MethodFunctionWithOptions`, try `locate det` or `code det`. It's more informative than producing nothing...

**Comment by mahrud** [comment]
I agree about `methodOptions`, but I would say the the behavior of `locate det` or `code det` is incorrect at the moment, and the reason is that `MethodFunctionWithOptions` is kind of a kludge at the moment.

**Comment by pzinn** [comment]
OK. let's just say, at some point in the future it'd be nice to have *some* information when one types `code <some method>` (even without `methods` inserted).

**Comment by mahrud** [comment]
I'm not disagreeing with you that they could do more, I'm just stating that `code` and `locate` are working as originally intended, which was to take a function closure object (say, `lookup(html, Net)`) and return some information stored inside that object.

However, the object `html` doesn't have any of that, so my question is what behavior do you wish they had? Do you want `code html` to be a synonym for `code methods html` or something else?

For `locate`, I really wish the location of, say, `html = method(...)` in `methods.m2` was returned.

**Comment by mahrud** [comment]
In other words, the reason `code` and `locate` return something for methods like `det`, `tensor`, or `hooks` (of types `MethodFunctionWithOptions`, `MethodFunctionBinary`, and `MethodFunctionSingle`, but what they all have in common is that they accept options) is that they are implemented as `FunctionClosure`s that they separate the optional arguments first, then do the method lookup based on the types of input and call the correct function.

However, methods that don't accept options do the whole lookup step in the interpreter (`evaluate.d`, I believe), so there is no code or location to return.

**Comment by pzinn** [comment]
> For locate, I really wish the location of, say, `html = method(...)` in `methods.m2` was returned.
yes precisely (and same for `det` once unkludged). an easy fix would be to have locate give the location of the symbol for non-functions. not perfect but would be something, since `locate symbol html` does the job ATM.

**Comment by mahrud** [comment]
> since `locate symbol html` does the job ATM.

Yes, that sounds great to me!

Btw, since you're working closely with code positions, would be great if you could improve the situation described here: https://github.com/Macaulay2/M2/issues/2569#issuecomment-1220868097

**Comment by pzinn** [comment]
ah, yes, these issues are related, I missed that thread.

**Comment by pzinn** [comment]
actually, using the location of the symbol (as in `locate symbol html`) doesn't quite work. I guess it does for protected symbols; otherwise the value of the symbol can be redefined, but its "location" won't (since the symbol only gets "created" the first time, I guess). so using `locate` on a symbol whose value has been redefined will still show the location of the original definition.

**Comment by pzinn** [comment]
there seems to be another problem with locating symbols. why is it that, say,
```
i2 : locate symbol det

o2 = mutablemat.m2:306:0-306:11

o2 : FilePosition
```
works fine, but
```
i3 : locate symbol resultant
```
(or trying to locate anything from a package, really) returns nothing??

**Comment by pzinn** [comment]
I just checked, your explanation is the correct one. That's sad.

**Comment by pzinn** [comment]
maybe this can be fixed. `export` calls `getGlobalSymbol` which creates a symbol with a `dummyPosition`. Maybe when a symbol is used and its position is dummy, its position can be updated?

**Comment by pzinn** [comment]
yeah that seems to work:
```
--- a/M2/Macaulay2/d/binding.d
+++ b/M2/Macaulay2/d/binding.d
@@ -436,6 +436,7 @@ lookup(t:Token,forcedef:bool,thread:bool):void := (
          when lookup(t.word,t.dictionary)
          is entry:Symbol do (
               t.entry = entry;
+              if entry.position == dummyPosition then entry.position = position(t);
               if entry.flagLookup then makeErrorTree(t,"flagged symbol encountered");
               if thread && !entry.thread then makeErrorTree(t,"symbol already present, but not thread local");
               )
```

**Comment by pzinn** [comment]
back to `methodOptions`: it seems the issue is that `MultipleArgsWithOptions` and `MultipleArgsNoOptions` have an argument `methopts` which is never actually used but allows recovering the options by using `frame`. why can't the exact same trick be used for `SingleWithOptions` and `SingleNoOptions`?

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Part of this landed and part is still live, so here is the state on 1.26.06.

**Fixed.** The complaint from 2023-03-03 that locating anything exported by a package returned
nothing — @pzinn's own patch for it went in as 1eb0011ded (2024-07-17, *"fixed locate for symbols
declared by export"*), and `binding.d` now carries the line he proposed here:

```
if entry.position == tempPosition then entry.position = t.position;
```

So `locate symbol resultant` gives `Elimination.m2:123` and `locate symbol html` gives
`methods.m2:713`.

**Still live.** `methodOptions html` is `null` — the one pzinn said should work no matter what — and
`locate html` is `null`. `code html` no longer returns nothing, but what it returns is
`method function with a single argument, html: source code not available`, which is a message rather
than an answer.

**Untouched, and probably the real ask.** Neither `locate` nor `code` points at the
`html = method(...)` line. On every method function I tried, across both classes in the title:

```
locate det          option.m2:15:14-18:8      (MethodFunctionWithOptions)
locate hooks        option.m2:15:14-18:8      (MethodFunctionSingle)
locate resultant    option.m2:15:14-18:8
```

all of them the generic wrapper — @mahrud's "kludge", and the same thing as #2569. And
`methodOptions hooks` is null too, even though `hooks` is the `MethodFunctionSingle` this issue is
titled for.

pzinn's last suggestion, from 2023-04-28 — that the `methopts` argument trick which lets
`MultipleArgsWithOptions` recover its options through `frame` could be applied to
`SingleWithOptions` and `SingleNoOptions` — was never answered, and looks like the shortest path to
the `methodOptions` half.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
